File: ObjC/HSSDKXObjcSample/HSSDKXObjcSample/ViewController.m
//
//  ViewController.m
//  HSSDKXObjcSample
//
//  Created by Sagar Dagdu on 24/05/21.
//  Copyright © 2021 Sagar Dagdu. All rights reserved.
//

#import "ViewController.h"
#import <HelpshiftX/HelpshiftX.h>

static NSString *const kHsDemoCellIdentifier = @"DemoCellIdentifier";

@interface ViewController ()<UITableViewDataSource, UITableViewDelegate, UITextFieldDelegate>

@property (weak, nonatomic) IBOutlet UITableView *actionsTableView;
@property (strong, nonatomic) NSArray<NSString *> *actionNames;

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];

    [self setupUI];
    self.actionNames = @[@"Show Conversation",
                         @"Show FAQs",
                         @"Show FAQ Section",
                         @"Show Single FAQ!",
                         @"Set Language"];
}

- (void)setupUI {
    self.title = @"SDK-X Objc Demo";
    [self.actionsTableView registerClass:[UITableViewCell class] forCellReuseIdentifier:kHsDemoCellIdentifier];
    self.actionsTableView.dataSource = self;
    self.actionsTableView.delegate = self;
}

- (nonnull UITableViewCell *)tableView:(nonnull UITableView *)tableView cellForRowAtIndexPath:(nonnull NSIndexPath *)indexPath {
    UITableViewCell *cell =  [self.actionsTableView dequeueReusableCellWithIdentifier:kHsDemoCellIdentifier forIndexPath:indexPath];

    cell.textLabel.text = [self.actionNames objectAtIndex:indexPath.row];
    return cell;
}

- (NSInteger)tableView:(nonnull UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return self.actionNames.count;
}

- (void) tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    [tableView deselectRowAtIndexPath:indexPath animated:YES];

    switch (indexPath.row) {
        case 0:
            [Helpshift showConversationWith:self config:nil];
            break;
        case 1:
            [Helpshift showFAQsWith:self config:nil];
        case 2:
#warning TODO: Replace the SECTION-ID with appropriate section id
            [Helpshift showFAQSection:@"SECTION-ID" with:self config:nil];
        case 3:
#warning TODO: Replace the FAQ-PUBLISH-ID with appropriate faq publish id
            [Helpshift showSingleFAQ:@"FAQ-PUBLISH-ID" with:self config:nil];
        case 4:
            [self openSetLanguageAlert];
            break;
        default:
            break;
    }
}

- (void)openSetLanguageAlert {
    UIAlertController *alert = [UIAlertController alertControllerWithTitle:@"Enter language code"
                                                                   message:nil
                                                            preferredStyle:UIAlertControllerStyleAlert];
    [alert addTextFieldWithConfigurationHandler:^(UITextField *textField) {
         textField.keyboardType = UIKeyboardTypeDefault;
         textField.delegate = self;
     }];
    
    UIAlertAction *yesAction = [UIAlertAction actionWithTitle:@"GO" style:UIAlertActionStyleDefault handler:^(UIAlertAction *action) {
                                    NSString *languageCode = alert.textFields.firstObject.text;
                                    [Helpshift setLanguage:languageCode];
                                }];
    UIAlertAction *noAction = [UIAlertAction actionWithTitle:@"Cancel" style:UIAlertActionStyleDestructive handler:^(UIAlertAction *action) {
                                   [alert dismissViewControllerAnimated:YES completion:nil];
                               }];
    [alert addAction:yesAction];
    [alert addAction:noAction];
    [self presentViewController:alert animated:YES completion:nil];
}

@end

File: Swift/HSSDKXSwiftSample/HSSDKXSwiftSample/ViewController.swift
//
//  ViewController.swift
//  HSSDKXSwiftSample
//
//  Created by Sagar Dagdu on 24/05/21.
//  Copyright © 2021 Sagar Dagdu. All rights reserved.
//

import UIKit
import HelpshiftX

fileprivate enum DemoCellAction: Int, CaseIterable {
    case showConversation
    case showFaqs
    case showFaqSection
    case showSingleFaq
    case setLanguage

    var displayTitle: String {
        get {
            switch self {
            case .showConversation:
                return "Show Conversation"
            case .showFaqs:
                return "Show Faqs"
            case .showFaqSection:
                return "Show FAQ Section"
            case .showSingleFaq:
                return "Show Single FAQ!"
            case .setLanguage:
                return "Set Language"
            }
        }
    }
}

class ViewController: UIViewController, UITextFieldDelegate {

    @IBOutlet private weak var actionsTableView: UITableView!
    
    private let demoCellIdentifier = "DemoCellIdentifier"
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        setupUI()
    }

    private func setupUI() {
        title = "SDK-X Swift Demo"
        actionsTableView.register(UITableViewCell.self, forCellReuseIdentifier: demoCellIdentifier)
        actionsTableView.dataSource = self
        actionsTableView.delegate = self
    }

    private func openSetLanguageAlert() {
        let alert = UIAlertController(title: "Enter language code", message: nil, preferredStyle: .alert)
        alert.addTextField { $0.delegate = self }
        let yesAction = UIAlertAction(title: "GO", style: .default) { (_) in
            if let languageCode = alert.textFields?.first?.text {
                Helpshift.setLanguage(languageCode)
            }
        }
        
        let noAction = UIAlertAction(title: "Cancel", style: .cancel) { (_) in
            alert.dismiss(animated: true, completion: nil)
        }
        
        alert.addAction(yesAction)
        alert.addAction(noAction)
        present(alert, animated: true, completion: nil)
    }
}

extension ViewController: UITableViewDataSource {
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return DemoCellAction.allCases.count
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = self.actionsTableView.dequeueReusableCell(withIdentifier: demoCellIdentifier, for: indexPath)
        cell.textLabel?.text = DemoCellAction(rawValue: indexPath.row)!.displayTitle
        return cell
    }
}

extension ViewController: UITableViewDelegate {
    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        tableView.deselectRow(at: indexPath, animated: true)
        
        guard let action = DemoCellAction(rawValue: indexPath.row) else { return }
        
        switch action {
        case .showConversation:
            Helpshift.showConversation(with: self, config: nil)
        case .showFaqs:
            Helpshift.showFAQs(with: self, config: nil)
        case .showFaqSection:
            #warning("TODO: Replace the SECTION-ID with appropriate section id")
            Helpshift.showFAQSection("SECTION-ID", with: self, config: nil)
        case .showSingleFaq:
            #warning("TODO: Replace the FAQ-PUBLISH-ID with appropriate faq publish id")
            Helpshift.showSingleFAQ("FAQ-PUBLISH-ID", with: self, config: nil)
        case .setLanguage:
            openSetLanguageAlert()
        }
        
    }
}
